sql/pipelines/src/main/scala/org/apache/spark/sql/pipelines/graph/Flow.scala

```diff
@@ -0,0 +1,230 @@
+   *
+   * @param allInputs the set of identifiers for all the [[Input]]s defined in the
+   *                  [[DataflowGraph]].
+   * @param availableInputs the list of all [[Input]]s available to this flow
+   * @param configuration the spark configurations that apply to this flow.
+   * @param currentCatalog The current catalog in execution context when the query is defined.
+   * @param currentDatabase The current database in execution context when the query is defined.
+   * @return the inputs actually used, and the [[DataFrame]] expression for the flow
+   */
+  def call(
+      allInputs: Set[TableIdentifier],
+      availableInputs: Seq[Input],
+      configuration: Map[String, String],
+      currentCatalog: Option[String],
+      currentDatabase: Option[String]
+  ): FlowFunctionResult
+}
+
+/**
+ * Holds the [[DataFrame]] returned by a [[FlowFunction]] along with the inputs used to
+ * construct it.
+ * @param usedBatchInputs the identifiers of the complete inputs read by the flow
+ * @param usedStreamingInputs the identifiers of the incremental inputs read by the flow
+ * @param usedExternalInputs the identifiers of the external inputs read by the flow
+ * @param dataFrame the [[DataFrame]] expression executed by the flow if the flow can be resolved
+ */
+case class FlowFunctionResult(
+    requestedInputs: Set[TableIdentifier],
+    usedBatchInputs: Set[ResolvedInput],
+    usedStreamingInputs: Set[ResolvedInput],
+    usedExternalInputs: Set[String],
+    dataFrame: Try[DataFrame],
+    sqlConf: Map[String, String],
+    analysisWarnings: Seq[AnalysisWarning] = Nil) {
+
+  /**
+   * Returns the names of all of the [[Input]]s used when resolving this [[Flow]]. If the
+   * flow failed to resolve, we return the all the datasets that were requested when evaluating the
+   * flow.
+   */
+  def inputs: Set[TableIdentifier] = {
+    (batchInputs ++ streamingInputs).map(_.input.identifier)
+  }
+
+  /** Names of [[Input]]s read completely by this [[Flow]]. */
+  def batchInputs: Set[ResolvedInput] = usedBatchInputs
+
+  /** Names of [[Input]]s read incrementally by this [[Flow]]. */
+  def streamingInputs: Set[ResolvedInput] = usedStreamingInputs
+
+  /** Returns errors that occurred when attempting to analyze this [[Flow]]. */
+  def failure: Seq[Throwable] = {
+    dataFrame.failed.toOption.toSeq
+  }
+
+  /** Whether this [[Flow]] is successfully analyzed. */
+  final def resolved: Boolean = failure.isEmpty // don't override this, override failure
+}
+
+/** A [[Flow]] whose output schema and dependencies aren't known. */
+class UnresolvedFlow(
+    val identifier: TableIdentifier,
+    val destinationIdentifier: TableIdentifier,
+    val func: FlowFunction,
+    val currentCatalog: Option[String],
+    val currentDatabase: Option[String],
+    val sqlConf: Map[String, String],
+    val comment: Option[String] = None,
+    override val once: Boolean,
+    override val origin: QueryOrigin
+) extends Flow {
+  def copy(
```

Comment:
shall we make this class a `case class` which provides `copy` method automatically?